File: web-app/frontend/lib/api/esrs.ts
/**
 * Server-only fetch for the ESRS-E1-aligned compliance report.
 * Types mirror backend app/schemas/esrs.py.
 */

export type EsrsScopeBreakdown = {
  scope1_tco2e: number
  scope2_location_tco2e: number
  scope2_market_tco2e: number
  scope3_tco2e: number
  total_location_tco2e: number
  total_market_tco2e: number
}

export type EsrsBuildingRow = {
  fabric_building_id: string
  name: string
  building_type: string
  floor_area_m2: number
  scope1_tco2e: number
  scope2_location_tco2e: number
  scope2_market_tco2e: number
  scope3_tco2e: number
  total_location_tco2e: number
  ghg_intensity_tco2e_m2: number | null
  data_quality_flag: string | null
}

export type EsrsDataQuality = {
  complete: number
  estimated: number
  missing_gas: number
  other: number
}

export type EsrsReport = {
  reporting_year: number | null
  has_data: boolean
  buildings_total: number
  buildings_reported: number
  floor_area_m2: number
  energy_total_mwh: number
  energy_renewable_pct: number | null
  ghg: EsrsScopeBreakdown
  ghg_intensity_tco2e_m2: number | null
  data_quality: EsrsDataQuality
  rows: EsrsBuildingRow[]
}

type FetchResult<T> = { ok: true; data: T } | { ok: false; error: string }

export async function fetchEsrsReport(
  accessToken: string,
  buildingId?: string | null
): Promise<FetchResult<EsrsReport>> {
  const backendUrl = process.env.BACKEND_URL || null
  if (!backendUrl) {
    console.error("[fetchEsrsReport] BACKEND_URL not configured")
    return { ok: false, error: "Server misconfigured" }
  }
  if (!accessToken) {
    return { ok: false, error: "Missing access token" }
  }

  const url = buildingId
    ? `${backendUrl}/compliance/esrs?building_id=${encodeURIComponent(buildingId)}`
    : `${backendUrl}/compliance/esrs`

  try {
    const res = await fetch(url, {
      method: "GET",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        Accept: "application/json",
      },
      cache: "no-store",
    })

    if (!res.ok) {
      const body = await res.text().catch(() => "<no body>")
      if (res.status === 503 && body.includes("fabric_unavailable")) {
        console.warn("[fetchEsrsReport] Fabric data temporarily unavailable")
        return { ok: false, error: "fabric_unavailable" }
      }
      console.error(`[fetchEsrsReport] Backend ${res.status}: ${body.slice(0, 200)}`)
      return { ok: false, error: `Backend ${res.status}` }
    }

    const data = (await res.json()) as EsrsReport
    return { ok: true, data }
  } catch (err) {
    console.error("[fetchEsrsReport] Network/unknown error:", err)
    return { ok: false, error: "Network error" }
  }
}

File: web-app/backend/app/schemas/esrs.py
"""Pydantic response models for the ESRS-E1-aligned compliance export.

ESRS-E1-aligned reporting SUPPORT — indicative, NOT an audited disclosure. GHG is
read from gold_ghg_scope (tCO2e, monthly per building per scope); energy from
gold_kpi_daily. Scope 3 is an estimate; Scope 2 is reported both location- and
market-based per ESRS E1-6.
"""
from pydantic import BaseModel, Field


class EsrsScopeBreakdown(BaseModel):
    """Portfolio gross GHG by scope for the reporting year (tCO2e)."""

    scope1_tco2e: float
    scope2_location_tco2e: float
    scope2_market_tco2e: float
    scope3_tco2e: float
    total_location_tco2e: float
    total_market_tco2e: float


class EsrsBuildingRow(BaseModel):
    """Per-building GHG breakdown for the reporting year."""

    fabric_building_id: str
    name: str
    building_type: str
    floor_area_m2: float
    scope1_tco2e: float
    scope2_location_tco2e: float
    scope2_market_tco2e: float
    scope3_tco2e: float
    total_location_tco2e: float
    ghg_intensity_tco2e_m2: float | None = None
    data_quality_flag: str | None = None


class EsrsDataQuality(BaseModel):
    """Count of reported buildings per data-quality flag (worst month wins)."""

    complete: int = 0
    estimated: int = 0
    missing_gas: int = 0
    other: int = 0


class EsrsReport(BaseModel):
    """ESRS-E1-aligned portfolio summary for one reporting year."""

    reporting_year: int | None = Field(
        default=None, description="Latest reporting year present in gold_ghg_scope."
    )
    reporting_months: int | None = Field(
        default=None, description="Months of data in the reporting year (<12 = partial/YTD)."
    )
    has_data: bool = False
    buildings_total: int = Field(default=0, description="Visible buildings.")
    buildings_reported: int = Field(
        default=0, description="Buildings with GHG data in the reporting year."
    )
    floor_area_m2: float = 0.0
    energy_total_mwh: float = 0.0
    energy_renewable_pct: float | None = Field(
        default=None,
        description="On-site solar self-consumed as a share of total consumption (%).",
    )
    ghg: EsrsScopeBreakdown
    ghg_intensity_tco2e_m2: float | None = Field(
        default=None, description="Total location-based GHG ÷ reported floor area."
    )
    data_quality: EsrsDataQuality
    rows: list[EsrsBuildingRow] = []
